### ISSUE TITLE
Steam: Borderlands 2 crashes when trying to download hotfix

### ISSUE THREAD

**Issue by bugworm** [NixOS/nixpkgs#25443]
## Issue description
Before entering the game it tries to download hotfix data, and crashes. I think this is it https://wiki.archlinux.org/index.php/Steam/Game-specific_troubleshooting#Logging_into_SHiFT but solutions from there doesn't work. Game works fine if launch it in offline mode

### Steps to reproduce
`nix-env -i steam`
Launch steam, log in, download borderlands 2 and launch it

## Technical details

* System: NixOS: 17.03.1082.4aab5c5798 (Gorilla)
* Nix version: nix-env (Nix) 1.11.8
* Nixpkgs version: 17.03.1082.4aab5c5798

**Comment by AnotherBugworm** [comment]
steam logs
```cp: cannot create regular file '/home/bugworm/.local/share/Steam/bootstrap.tar.xz': Permission denied
Running Steam on nixos 17.03.1082.4aab5c5798 64-bit
STEAM_RUNTIME has been set by the user to: /steamrt
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Fontconfig error: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 70: non-double matrix element
Fontconfig error: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 70: non-double matrix element
Fontconfig warning: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 78: saw unknown, expected number
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)
Installing breakpad exception handler for appid(steam)/version(1493162727)

** (steam:2794): WARNING **: Unknown device type 14

** (steam:2794): WARNING **: Could not create object for /org/freedesktop/NetworkManager/Devices/0: unknown object type

** (steam:2794): WARNING **: Ignoring invalid property 'secondaries'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'secondaries'

** (steam:2794): WARNING **: Ignoring invalid property 'autoconnect-priority'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'secondaries'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'secondaries'

** (steam:2794): WARNING **: Ignoring invalid property 'autoconnect-priority'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'

** (steam:2794): WARNING **: Ignoring invalid property 'route-data'

** (steam:2794): WARNING **: Ignoring invalid property 'address-data'
Installing breakpad exception handler for appid(steam)/version(1493162727)
Generating new string page texture 2: 48x256, total string texture memory is 49.15 KB
Generating new string page texture 3: 256x256, total string texture memory is 311.30 KB
Installing breakpad exception handler for appid(steam)/version(1493162727)
/run/current-system/sw/share/themes/Breeze-Dark/gtk-2.0/widgets/entry:70: error: unexpected identifier `direction', expected character `}'

(steam:2794): Gtk-WARNING **: Unable to locate theme engine in module_path: "adwaita",
/run/current-system/sw/share/themes/Breeze-Dark/gtk-2.0/widgets/styles:36: error: invalid string constant "combobox_entry", expected valid string constant
Installing breakpad exception handler for appid(steam)/version(1493162727)
roaming config store loaded successfully - 1323 bytes.
migrating temporary roaming config store
Installing breakpad exception handler for appid(steam)/version(1493162727)
Fontconfig error: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 70: non-double matrix element
Fontconfig error: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 70: non-double matrix element
Fontconfig warning: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 78: saw unknown, expected number
Failed to init SteamVR because it isn't installed
ExecCommandLine: ""/home/bugworm/.local/share/Steam/ubuntu12_32/steam" "
Installing breakpad exception handler for appid(steam)/version(1493162727)
System startup time: 4.44 seconds
Generating new string page texture 74: 1024x256, total string texture memory is 1.36 MB
Generating new string page texture 75: 128x256, total string texture memory is 131.07 KB
Generating new string page texture 76: 128x256, total string texture memory is 1.49 MB
Generating new string page texture 77: 64x256, total string texture memory is 1.56 MB
Generating new string page texture 78: 8x256, total string texture memory is 1.56 MB
Generating new string page texture 79: 32x256, total string texture memory is 1.60 MB

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.
Generating new string page texture 88: 128x256, total string texture memory is 1.73 MB
Generating new string page texture 90: 256x256, total string texture memory is 1.99 MB
Generating new string page texture 118: 384x256, total string texture memory is 2.38 MB
Game update: AppID 49520 "Borderlands 2", ProcID 2856, IP 0.0.0.0:0
>>> Adding process 2856 for game ID 49520
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_64/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
/home/bugworm/.local/share/Steam/steamapps/common/Borderlands 2/./Borderlands2: /steamrt/i386/lib/i386-linux-gnu/libdbus-1.so.3: no version information available (required by /steamrt/i386/lib/libpulse.so.0)
/home/bugworm/.local/share/Steam/steamapps/common/Borderlands 2/./Borderlands2: /steamrt/i386/lib/i386-linux-gnu/libdbus-1.so.3: no version information available (required by /nix/store/3h9x3ya9cbwm53hjj177gwzfszxbibkj-libpulseaudio-10.0/lib/pulseaudio/libpulsecommon-10.0.so)
>>> Adding process 2857 for game ID 49520

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

(steam:2794): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.
Setting breakpad minidump AppID = 49520
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561198040972398 [API loaded no]
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
lspci: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
>>> Adding process 2859 for game ID 49520
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/bugworm/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
lspci: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
>>> Adding process 2866 for game ID 49520
Installing breakpad exception handler for appid(gameoverlayui)/version(20170425224519)
Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)
Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)
Fontconfig error: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 70: non-double matrix element
Fontconfig error: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 70: non-double matrix element
Fontconfig warning: "/etc/fonts//conf.d/10-scale-bitmap-fonts.conf", line 78: saw unknown, expected number
Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)
Generating new string page texture 132: 256x256, total string texture memory is 393.22 KB
Generating new string page texture 133: 128x256, total string texture memory is 2.51 MB
Game removed: AppID 49520 "Borderlands 2", ProcID 2857 
No cached sticky mapping in ActivateActionSet.
```

**Comment by stanipintjuk** [comment]
I have the same problem.

I have done some investigating and these are my results:

* ```ldd``` doesn't complain about any missing libraries.

```
./Borderlands2_bak: /steamrt/i386/lib/i386-linux-gnu/libdbus-1.so.3: no version information available (required by /nix/store/6ypyfg3r421vnxxxiyxq4x23vnqxjzfh-libpulseaudio-10.0/lib/pulseaudio/libpulsecommon-10.0.so)
        linux-gate.so.1 (0xf77a2000)
        /home/stani/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so (0xf7768000)
        libsteam_api.so => ./libsteam_api.so (0xf7758000)
        libcrypto.so.1.0.0 => ./libcrypto.so.1.0.0 (0xf75ab000)
        libX11.so.6 => /steamrt/i386/usr/lib/i386-linux-gnu/libX11.so.6 (0xf7477000)
        libGL.so.1 => /run/opengl-driver-32/lib/libGL.so.1 (0xf73cb000)
        libz.so.1 => /steamrt/i386/lib/i386-linux-gnu/libz.so.1 (0xf73b5000)
        libpthread.so.0 => /usr/lib32/libpthread.so.0 (0xf7397000)
        librt.so.1 => /usr/lib32/librt.so.1 (0xf738e000)
        libdl.so.2 => /usr/lib32/libdl.so.2 (0xf7389000)
        libm.so.6 => /usr/lib32/libm.so.6 (0xf732e000)
        libuuid.so.1 => /steamrt/i386/lib/i386-linux-gnu/libuuid.so.1 (0xf7328000)
        libc++.so.1 => ./libc++.so.1 (0xf727e000)
        libcxxrt.so => ./libcxxrt.so (0xf725f000)
        libc.so.6 => /usr/lib32/libc.so.6 (0xf70a5000)
        libgcc_s.so.1 => /steamrt/i386/lib/i386-linux-gnu/libgcc_s.so.1 (0xf708a000)
        libiconv.so.2 => ./libiconv.so.2 (0xf6fa6000)
        libstdc++.so.6 => /steamrt/i386/usr/lib/i386-linux-gnu/libstdc++.so.6 (0xf6ebd000)
        libssl.so.1.0.0 => ./libssl.so.1.0.0 (0xf6e67000)
        libopenal.so.1 => /steamrt/i386/lib/libopenal.so.1 (0xf6dc8000)
        /lib/ld-linux.so.2 (0xf77a4000)
        libxcb.so.1 => /steamrt/i386/usr/lib/i386-linux-gnu/libxcb.so.1 (0xf6da6000)
        libXext.so.6 => /steamrt/i386/usr/lib/i386-linux-gnu/libXext.so.6 (0xf6d94000)
        libasound.so.2 => /steamrt/i386/lib/libasound.so.2 (0xf6c72000)
        libpulse.so.0 => /steamrt/i386/lib/libpulse.so.0 (0xf6c17000)
        libXau.so.6 => /steamrt/i386/usr/lib/i386-linux-gnu/libXau.so.6 (0xf6c11000)
        libXdmcp.so.6 => /steamrt/i386/usr/lib/i386-linux-gnu/libXdmcp.so.6 (0xf6c0a000)
        libpulsecommon-10.0.so => /nix/store/6ypyfg3r421vnxxxiyxq4x23vnqxjzfh-libpulseaudio-10.0/lib/pulseaudio/libpulsecommon-10.0.so (0xf6b7d000)
        libsndfile.so.1 => /steamrt/i386/usr/lib/i386-linux-gnu/libsndfile.so.1 (0xf6b0b000)
        libFLAC.so.8 => /steamrt/i386/usr/lib/i386-linux-gnu/libFLAC.so.8 (0xf6abd000)
        libvorbisenc.so.2 => /steamrt/i386/usr/lib/i386-linux-gnu/libvorbisenc.so.2 (0xf6945000)
        libvorbis.so.0 => /steamrt/i386/usr/lib/i386-linux-gnu/libvorbis.so.0 (0xf691a000)
        libogg.so.0 => /steamrt/i386/usr/lib/i386-linux-gnu/libogg.so.0 (0xf6910000)
        libdbus-1.so.3 => /steamrt/i386/lib/i386-linux-gnu/libdbus-1.so.3 (0xf68c6000)
        libsystemd.so.0 => /nix/store/s4dd533qfvbrnjh8ryyzhq45m0h2kxhp-systemd-232-lib/lib/libsystemd.so.0 (0xf6836000)
        libcap.so.2 => /steamrt/i386/lib/i386-linux-gnu/libcap.so.2 (0xf6830000)
        libresolv.so.2 => /usr/lib32/libresolv.so.2 (0xf6819000)
        liblzma.so.5 => /nix/store/cijfycgfilg76p91m37blvz0m27pchig-xz-5.2.3/lib/liblzma.so.5 (0xf67ee000)
        liblz4.so.1 => /nix/store/ndv58ld00rj8f8glyax0hbixjamrnyz9-lz4-131/lib/liblz4.so.1 (0xf67db000)
        libgcrypt.so.20 => /nix/store/3hzinps6j8hgj3wcxf1xr6cjz21x8dhb-libgcrypt-1.7.6/lib/libgcrypt.so.20 (0xf670d000)
        libgpg-error.so.0 => /steamrt/i386/lib/libgpg-error.so.0 (0xf66f3000)
```

* Running Borderlands 2 through steam with the 64bit libraries removed from ```LD_LIBRARY_PATH```
it instead complains about libGL being the wrong class

```
Setting breakpad minidump AppID = 49520
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561198188863205 [API loaded no]
ERROR: ld.so: object '/home/stani/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
sh: error while loading shared libraries: libGL.so.1: wrong ELF class: ELFCLASS32
ERROR: ld.so: object '/home/stani/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
sh: error while loading shared libraries: libGL.so.1: wrong ELF class: ELFCLASS32
```

Seems that it needs the 64bit libGL, but why would a 32bit game want to use a 64bit lib?

**Comment by stanipintjuk** [comment]
I managed to make the game run by turning off my internet connect right before it launches so it doesn't start downloading the hotfix.

When it runs; it runs smoothly, but still complains about libGL.so.1 in the output, which leads me to believe that there is nothing wrong with that lib and I have been looking in the wrong place -.-

**Comment by Mounium** [comment]
For me it only works if I start steam in offline mode..

**Comment by orivej** [comment]
@DIzFer has fixed this in #31480
